lib/compiler/scjson-to-module.js

```diff
@@ -183,142 +188,200 @@ function generateModule(rootState, moduleType){
-            throw new Error('AMD not yet supported');
-            break;
-        case 'umd':
-            throw new Error('UMD not yet supported');
-            break;
-        case 'commonjs':
-            r = 'module.exports = ' + s + ';';
-            break;
-        default:
-            r = s;
-            break;
-    }
 
-    r = o.headerString + '\n' + r;
+    var s =
+        o.headerString + '\n' +
+        generateFactoryFunctionWrapper(o, rootState.name, options);
 
-    return r;
+    return s;
 }
 
 function markAsReference(fnName){
     return  REFERENCE_MARKER + fnName + REFERENCE_MARKER;
 }
 
-function replaceActions(actionContainer,actionPropertyName){
+ModuleBuilder.prototype.replaceActions = function(actionContainer,actionPropertyName){
     if(actionContainer[actionPropertyName]){
         var actions = Array.isArray(actionContainer[actionPropertyName]) ? actionContainer[actionPropertyName] : [actionContainer[actionPropertyName]] ;
-        
-        actionContainer[actionPropertyName] = actions.map(generateActionFunction).map(markAsReference);
+
+        actionContainer[actionPropertyName] = actions.map(this.generateActionFunction, this).map(markAsReference);
 
         if(actionContainer[actionPropertyName].length === 1){
             actionContainer[actionPropertyName] = actionContainer[actionPropertyName][0];
         }
     }
 }
 
-function visitState(state){
-    //accumulate datamodels
-    if(state.datamodel){
-        datamodelAccumulator.push.apply(datamodelAccumulator,state.datamodel);
+ModuleBuilder.prototype.visitState = function(){
+    var genValue = this.stateGen.next();
+    if (genValue.done) {
+        this._finish();
+        return;
     }
 
-    if(state.onExit) replaceActions(state,'onExit');
-    if(state.onEntry) replaceActions(state,'onEntry');
+    var state = genValue.value;
+    //accumulate datamodels
+    if (state.datamodel) {
+        this.datamodelAccumulator.push.apply(this.datamodelAccumulator,state.datamodel);
+    }
 
-    if(state.transitions){
-        state.transitions.forEach(function(transition){
-            replaceActions(transition,'onTransition');
+    if(state.onExit) this.replaceActions(state,'onExit');
+    if(state.onEntry) this.replaceActions(state,'onEntry');
 
+    if(state.transitions) {
+        for (var i = 0, len = state.transitions.length; i < len; i++) {
+            var transition = state.transitions[i];
+            this.replaceActions(transition,'onTransition');
             if(transition.cond){
-                transition.cond = markAsReference(generateAttributeExpression(transition,'cond'));
+                transition.cond = markAsReference(this.generateAttributeExpression(transition,'cond'));
             }
-        });
+        }
     }
 
     //clean up as we go
     delete state.datamodel;
 
-    if(state.states) state.states.forEach(function(substate){ visitState(substate); });
+    setImmediate(function(self) {
+        self.visitState();
+    }, this);
```

Comment:
Why make this asynchronous? Is it because visitState is expensive for large models?